### PR TITLE
Support for multi-device provisioning flow.

### PR DESCRIPTION
**Pull request by moxie0**
// FREEBIE

### CODE REVIEW

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java [hunk 1]
```diff
@@ -79,18 +79,22 @@
...
+  private static final String PROVISIONING_MESSAGE_PATH = "/v1/messages/provisioning/%s";
+
   private static final String DIRECTORY_TOKENS_PATH     = "/v1/directory/tokens";
   private static final String DIRECTORY_VERIFY_PATH     = "/v1/directory/%s";
   private static final String MESSAGE_PATH              = "/v1/messages/%s";
   private static final String RECEIPT_PATH              = "/v1/receipt/%s/%d";
   private static final String ATTACHMENT_PATH           = "/v1/attachments/%s";
 
-  private static final boolean ENFORCE_SSL = true;
+  private static final boolean ENFORCE_SSL = false;
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java]: ??

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/crypto/ProvisioningCipher.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  public byte[] encrypt(ProvisionMessage message) throws InvalidKeyException {
+    ECKeyPair ourKeyPair    = Curve.generateKeyPair();
+    byte[]    sharedSecret  = Curve.calculateAgreement(theirPublicKey, ourKeyPair.getPrivateKey());
+    byte[]    derivedSecret = new HKDFv3().deriveSecrets(sharedSecret, theirPublicKey.serialize(), "TextSecure Provisioning Message".getBytes(), 64);
+    byte[][]  parts         = Util.split(derivedSecret, 32, 32);
+
+    byte[] version    = {0x01};
+    byte[] ciphertext = getCiphertext(parts[0], message.toByteArray());
+    byte[] mac        = getMac(parts[1], Util.join(version, ciphertext));
+    byte[] body =  Util.join(version, ciphertext, mac);
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/crypto/ProvisioningCipher.java]: space nitpick

File: res/layout/device_provisioning_activity.xml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                  android:textAppearance="?android:attr/textAppearanceLarge"
+                  android:gravity="center"
+                  android:text="Add device?"
+                  android:id="@+id/textView"/>
+
+        <TextView android:layout_width="wrap_content"
+                  android:layout_height="wrap_content"
+                  android:textAppearance="?android:attr/textAppearanceMedium"
+                  android:gravity="left"
+                  android:text="The device you scanned will receive access to all incoming and outgoing messages."/>
```
Comment (mcginty) [inline res/layout/device_provisioning_activity.xml]: localization

File: res/layout/device_provisioning_activity.xml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    <LinearLayout android:layout_width="wrap_content"
+                  android:layout_height="wrap_content"
+                  android:orientation="vertical"
+                  android:layout_centerInParent="true">
+
+        <TextView android:layout_width="fill_parent"
+                  android:layout_height="wrap_content"
+                  android:textAppearance="?android:attr/textAppearanceLarge"
+                  android:gravity="center"
+                  android:text="Add device?"
```
Comment (mcginty) [inline res/layout/device_provisioning_activity.xml]: localization

File: res/layout/device_provisioning_activity.xml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        <LinearLayout android:layout_width="wrap_content"
+                      android:layout_gravity="center"
+                      android:layout_height="wrap_content"
+                      android:orientation="horizontal">
+
+            <!--<ProgressBar android:id="@+id/progress"-->
+                         <!--android:layout_width="wrap_content"-->
+                         <!--android:layout_height="wrap_content"-->
+                         <!--android:indeterminate="true"-->
+                         <!--android:visibility="gone"/>-->
```
Comment (mcginty) [inline res/layout/device_provisioning_activity.xml]: commented out stuff probably unneeded

File: res/layout/device_provisioning_activity.xml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+            <Button android:id="@+id/cancel_button"
+                    android:layout_width="0dip"
+                    android:layout_weight="1"
+                    android:layout_height="wrap_content"
+                    android:text="Cancel"/>
```
Comment (mcginty) [inline res/layout/device_provisioning_activity.xml]: R.string

File: res/layout/device_provisioning_activity.xml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+            <Button android:id="@+id/continue_button"
+                    android:layout_width="0dip"
+                    android:layout_weight="1"
+                    android:layout_height="wrap_content"
+                    android:text="Continue"/>
```
Comment (mcginty) [inline res/layout/device_provisioning_activity.xml]: R.string

File: src/org/thoughtcrime/securesms/DeviceProvisioningActivity.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    this.cancelButton.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        finish();
+      }
+    });
+  }
+
+  private void handleProvisioning() {
+    new ProgressDialogAsyncTask<Void, Void, Integer>(this, "Adding device...", "Adding new device...") {
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/DeviceProvisioningActivity.java]: R.string

File: src/org/thoughtcrime/securesms/DeviceProvisioningActivity.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            finish();
+            break;
+          case NO_DEVICE:
+            Toast.makeText(context, "No device found!", Toast.LENGTH_LONG).show();
+            break;
+          case NETWORK_ERROR:
+            Toast.makeText(context, "Network error!", Toast.LENGTH_LONG).show();
+            break;
+          case KEY_ERROR:
+            Toast.makeText(context, "Invalid QR code!", Toast.LENGTH_LONG).show();
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/DeviceProvisioningActivity.java]: R.string for alla'these too. If you don't want to bother I can just use these R.string comments as notes for when I "materialize" it.

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java (hunk 1 above)
Comment (FredericJacobs) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java]: I was about to report that too :+1:

File: res/layout/device_provisioning_activity.xml (hunk 3 above)
Comment (McLoo) [inline res/layout/device_provisioning_activity.xml]: ...access to all incoming and outgoing *push(?)* messages *except for self sends(?)*

Comment (moxie0) [inline res/layout/device_provisioning_activity.xml]: @mcginty I don't stringify stuff i figure you're going to restyle.

**Comment by mcginty** [comment]
@moxie0 want to check out https://github.com/WhisperSystems/TextSecure/commit/9168f0e4f7ed92f5c12a149a9d3cc66074c108d7 and if you're cool with it we can merge this?

File: src/org/thoughtcrime/securesms/DeviceProvisioningActivity.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    super.onCreate(bundle);
+    getSupportActionBar().hide();
+    initializeResources();
+
+    SpannableStringBuilder content = new SpannableStringBuilder();
+    content.append(getString(R.string.DeviceProvisioning_content_intro))
+           .append("\n")
+           .append(small(getString(R.string.DeviceProvisioning_content_bullets)));
+
+    new Builder(this).title("Link this device?")
```
Comment (McLoo) [inline src/org/thoughtcrime/securesms/DeviceProvisioningActivity.java]: L10n

**Pull request closed without merging**